File: DotNet/Communications/IridiumSatellite.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Windows.Forms;
using AGI.Foundation.Celestial;
using AGI.Foundation.Communications;
using AGI.Foundation.Communications.Antennas;
using AGI.Foundation.Communications.SignalProcessing;
using AGI.Foundation.Coordinates;
using AGI.Foundation.Geometry;
using AGI.Foundation.Graphics;
using AGI.Foundation.Graphics.Advanced;
using AGI.Foundation.Graphics.Renderer;
using AGI.Foundation.Platforms;
using AGI.Foundation.Propagators;

namespace Communications
{
    /// <summary>
    /// A simple Platform based class which models an Iridium satellite.
    /// It also adds a Transceiver property as well as default extensions for graphics.
    /// </summary>
    public class IridiumSatellite : Platform
    {
        /// <summary>
        /// Creates a new instance with the provided properties.
        /// </summary>
        public IridiumSatellite(string name, IEnumerable<TwoLineElementSet> tles, Font font, double targetFrequency)
        {
            var earth = CentralBodiesFacet.GetFromContext().Earth;

            //Set the name and create the point from the TLE.
            Name = name;
            LocationPoint = new Sgp4Propagator(tles).CreatePoint();
            OrientationAxes = new AxesVehicleVelocityLocalHorizontal(earth.InertialFrame, LocationPoint);

            //Create a transceiver modeled after Iridium specs.
            m_transceiver = new Transceiver
            {
                Name = Name + " Transceiver",
                Modulation = new ModulationQpsk(),
                OutputGain = 100.0,
                OutputNoiseFactor = 1.2,
                CarrierFrequency = targetFrequency,
                Filter = new RectangularFilter
                {
                    Frequency = targetFrequency,
                    UpperBandwidthLimit = 20.0e6,
                    LowerBandwidthLimit = -20.0e6
                },
                InputAntennaGainPattern = new GaussianGainPattern(1.0, 0.55, 0.001),
                OutputAntennaGainPattern = new GaussianGainPattern(1.0, 0.55, 0.001)
            };
            m_transceiver.InputAntenna.LocationPoint = new PointFixedOffset(ReferenceFrame, new Cartesian(0, -1.0, 0));
            m_transceiver.OutputAntenna.LocationPoint = new PointFixedOffset(ReferenceFrame, new Cartesian(0, +1.0, 0));

            // setup Marker graphics for the satellites
            var satelliteTexture = SceneManager.Textures.FromUri(Path.Combine(Application.StartupPath, "Data/Markers/smallsatellite.png"));

            m_markerGraphicsExtension = new MarkerGraphicsExtension(new MarkerGraphics
            {
                Texture = new ConstantGraphicsParameter<Texture2D>(satelliteTexture),
                DisplayParameters = new DisplayParameters
                {
                    // hide marker when camera is closer than the below distance
                    MinimumDistance = new ConstantGraphicsParameter<double>(17500000.0)
                }
            });
            Extensions.Add(m_markerGraphicsExtension);

            // setup Model graphics for the satellites
            var modelUri = new Uri(Path.Combine(Application.StartupPath, "Data/Models/iridium.mdl"));

            m_modelGraphicsExtension = new ModelGraphicsExtension(new ModelGraphics
            {
                Uri = new ConstantGraphicsParameter<Uri>(modelUri),
                Scale = new ConstantGraphicsParameter<double>(50000.0),
                DisplayParameters = new DisplayParameters
                {
                    // hide model when camera is further than the marker minimum distance above
                    MaximumDistance = m_markerGraphicsExtension.MarkerGraphics.DisplayParameters.MinimumDistance
                }
            });
            Extensions.Add(m_modelGraphicsExtension);

            // setup text graphics for the satellites
            m_textGraphicsExtension = new TextGraphicsExtension(new TextGraphics
            {
                Color = new ConstantGraphicsParameter<Color>(Color.Red),
                Font = new ConstantGraphicsParameter<Font>(font),
                Outline = new ConstantGraphicsParameter<bool>(true),
                OutlineColor = new ConstantGraphicsParameter<Color>(Color.Black),
                Text = new ConstantGraphicsParameter<string>(Name),
                Origin = new ConstantGraphicsParameter<Origin>(Origin.TopCenter),
                PixelOffset = new ConstantGraphicsParameter<PointF>(new PointF(0, -satelliteTexture.Template.Height / 2))
            });
            if (TextureFilter2D.Supported(TextureWrap.ClampToEdge))
            {
                m_textGraphicsExtension.TextGraphics.TextureFilter = new ConstantGraphicsParameter<TextureFilter2D>(TextureFilter2D.NearestClampToEdge);
            }

            Extensions.Add(m_textGraphicsExtension);
        }

        /// <summary>
        /// Gets the Transceiver on this satellite.
        /// </summary>
        public Transceiver Transceiver
        {
            get { return m_transceiver; }
        }

        /// <summary>
        /// Gets the marker graphics.
        /// </summary>
        public MarkerGraphics MarkerGraphics
        {
            get { return m_markerGraphicsExtension.MarkerGraphics; }
        }

        /// <summary>
        /// Gets the model graphics.
        /// </summary>
        public ModelGraphics ModelGraphics
        {
            get { return m_modelGraphicsExtension.ModelGraphics; }
        }

        /// <summary>
        /// Gets the text graphics.
        /// </summary>
        public TextGraphics TextGraphics
        {
            get { return m_textGraphicsExtension.TextGraphics; }
        }

        private readonly TextGraphicsExtension m_textGraphicsExtension;
        private readonly MarkerGraphicsExtension m_markerGraphicsExtension;
        private readonly ModelGraphicsExtension m_modelGraphicsExtension;
        private readonly Transceiver m_transceiver;
    }
}

File: Java/CommunicationsDemo/src/agi/examples/communicationsdemo/IridiumSatellite.java
package agi.examples.communicationsdemo;

import java.awt.Color;
import java.awt.Font;
import java.io.File;
import java.net.URI;

import agi.foundation.celestial.CentralBodiesFacet;
import agi.foundation.celestial.EarthCentralBody;
import agi.foundation.communications.Transceiver;
import agi.foundation.communications.antennas.GaussianGainPattern;
import agi.foundation.communications.signalprocessing.ModulationQpsk;
import agi.foundation.communications.signalprocessing.RectangularFilter;
import agi.foundation.compatibility.PointF;
import agi.foundation.coordinates.Cartesian;
import agi.foundation.geometry.AxesVehicleVelocityLocalHorizontal;
import agi.foundation.geometry.PointFixedOffset;
import agi.foundation.graphics.SceneManager;
import agi.foundation.graphics.advanced.Origin;
import agi.foundation.graphics.renderer.Texture2D;
import agi.foundation.graphics.renderer.TextureFilter2D;
import agi.foundation.graphics.renderer.TextureWrap;
import agi.foundation.platforms.ConstantGraphicsParameter;
import agi.foundation.platforms.MarkerGraphics;
import agi.foundation.platforms.MarkerGraphicsExtension;
import agi.foundation.platforms.ModelGraphics;
import agi.foundation.platforms.ModelGraphicsExtension;
import agi.foundation.platforms.Platform;
import agi.foundation.platforms.TextGraphics;
import agi.foundation.platforms.TextGraphicsExtension;
import agi.foundation.propagators.Sgp4Propagator;
import agi.foundation.propagators.TwoLineElementSet;

/**
 * A simple Platform based class which models an Iridium satellite.
 * It also adds a Transceiver property as well as default extensions for graphics.
 */
public class IridiumSatellite extends Platform {
    /**
     * Creates a new instance with the provided properties.
     */
    public IridiumSatellite(String name, Iterable<TwoLineElementSet> tles, Font font, double targetFrequency) {
        EarthCentralBody earth = CentralBodiesFacet.getFromContext().getEarth();

        // Set the name and create the point from the TLE.
        setName(name);
        setLocationPoint(new Sgp4Propagator(tles).createPoint());
        setOrientationAxes(new AxesVehicleVelocityLocalHorizontal(earth.getInertialFrame(), getLocationPoint()));

        // Create a transceiver modeled after Iridium specs.
        m_transceiver = new Transceiver();
        m_transceiver.setName(getName() + " Transceiver");
        m_transceiver.setModulation(new ModulationQpsk());
        m_transceiver.setOutputGain(100.0);
        m_transceiver.setOutputNoiseFactor(1.2);
        m_transceiver.setCarrierFrequency(targetFrequency);
        RectangularFilter filter = new RectangularFilter();
        filter.setFrequency(targetFrequency);
        filter.setUpperBandwidthLimit(20.0e6);
        filter.setLowerBandwidthLimit(-20.0e6);
        m_transceiver.setFilter(filter);
        m_transceiver.setInputAntennaGainPattern(new GaussianGainPattern(1.0, 0.55, 0.001));
        m_transceiver.setOutputAntennaGainPattern(new GaussianGainPattern(1.0, 0.55, 0.001));
        m_transceiver.getInputAntenna().setLocationPoint(new PointFixedOffset(getReferenceFrame(), new Cartesian(0, -1.0, 0)));
        m_transceiver.getOutputAntenna().setLocationPoint(new PointFixedOffset(getReferenceFrame(), new Cartesian(0, +1.0, 0)));

        // setup Marker graphics for the satellites
        Texture2D satelliteTexture = SceneManager.getTextures().fromUri("Data/Markers/smallsatellite.png");

        MarkerGraphics markerGraphics = new MarkerGraphics();
        markerGraphics.setTexture(new ConstantGraphicsParameter<>(satelliteTexture));
        // hide marker when camera is closer than the below distance
        markerGraphics.getDisplayParameters().setMinimumDistance(new ConstantGraphicsParameter<>(17500000.0));

        m_markerGraphicsExtension = new MarkerGraphicsExtension(markerGraphics);
        getExtensions().add(m_markerGraphicsExtension);

        // setup Model graphics for the satellites
        URI modelUri = new File("Data/Models/iridium.mdl").toURI();

        ModelGraphics modelGraphics = new ModelGraphics();
        modelGraphics.setUri(new ConstantGraphicsParameter<>(modelUri));
        modelGraphics.setScale(new ConstantGraphicsParameter<>(50000.0));
        // hide model when camera is further than the marker minimum distance above
        modelGraphics.getDisplayParameters().setMaximumDistance(markerGraphics.getDisplayParameters().getMinimumDistance());

        m_modelGraphicsExtension = new ModelGraphicsExtension(modelGraphics);
        getExtensions().add(m_modelGraphicsExtension);

        // setup text graphics for the satellites
        TextGraphics textGraphics = new TextGraphics();
        textGraphics.setColor(new ConstantGraphicsParameter<>(Color.RED));
        textGraphics.setFont(new ConstantGraphicsParameter<>(font));
        textGraphics.setOutline(new ConstantGraphicsParameter<>(true));
        textGraphics.setOutlineColor(new ConstantGraphicsParameter<>(Color.BLACK));
        textGraphics.setText(new ConstantGraphicsParameter<>(getName()));
        textGraphics.setOrigin(new ConstantGraphicsParameter<>(Origin.TOP_CENTER));
        textGraphics.setPixelOffset(new ConstantGraphicsParameter<>(new PointF(0, -satelliteTexture.getTemplate().getHeight() / 2)));
        if (TextureFilter2D.supported(TextureWrap.CLAMP_TO_EDGE)) {
            textGraphics.setTextureFilter(new ConstantGraphicsParameter<>(TextureFilter2D.getNearestClampToEdge()));
        }

        m_textGraphicsExtension = new TextGraphicsExtension(textGraphics);
        getExtensions().add(m_textGraphicsExtension);
    }

    // Gets the Transceiver on this satellite.
    public Transceiver getTransceiver() {
        return m_transceiver;
    }

    // Gets the marker graphics.
    public MarkerGraphics getMarkerGraphics() {
        return m_markerGraphicsExtension.getMarkerGraphics();
    }

    // Gets the model graphics.
    public ModelGraphics getModelGraphics() {
        return m_modelGraphicsExtension.getModelGraphics();
    }

    // Gets the text graphics.
    public TextGraphics getTextGraphics() {
        return m_textGraphicsExtension.getTextGraphics();
    }

    private final TextGraphicsExtension m_textGraphicsExtension;
    private final MarkerGraphicsExtension m_markerGraphicsExtension;
    private final ModelGraphicsExtension m_modelGraphicsExtension;
    private final Transceiver m_transceiver;
}
